src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -0,0 +1,131 @@
+#include "gcs_node_manager.h"
+#include <ray/common/ray_config.h>
+#include <ray/gcs/pb_util.h>
+#include <ray/rpc/node_manager/node_manager_client.h>
+#include "ray/gcs/redis_gcs_client.h"
+
+namespace ray {
+namespace gcs {
+
+GcsNodeManager::GcsNodeManager(boost::asio::io_service &io_service,
+                               std::shared_ptr<gcs::RedisGcsClient> gcs_client)
+    : client_call_manager_(io_service),
+      gcs_client_(std::move(gcs_client)),
+      num_heartbeats_timeout_(RayConfig::instance().num_heartbeats_timeout()),
+      heartbeat_timer_(io_service) {}
+
+void GcsNodeManager::HandleHeartbeat(const ClientID &node_id,
+                                     const rpc::HeartbeatTableData &heartbeat_data) {
+  heartbeats_[node_id] = num_heartbeats_timeout_;
+  heartbeat_buffer_[node_id] = heartbeat_data;
+}
+
+void GcsNodeManager::Start() {
+  RAY_LOG(INFO) << "Starting gcs node manager.";
+  const auto heartbeat_callback = [this](const ClientID &id,
+                                         const rpc::HeartbeatTableData &heartbeat_data) {
+    HandleHeartbeat(id, heartbeat_data);
+  };
+  RAY_CHECK_OK(gcs_client_->Nodes().AsyncSubscribeHeartbeat(heartbeat_callback, nullptr));
```

Comment:
I think there is no need to subscribe to redis for heartbeats. Because raylets will send heartbeats to GCS Server first, just invoke `HandleHeartbeat` when receiving a heartbeat request.